### PR TITLE
[FLINK-26179] Support for periodic savepoints

### PR DESCRIPTION
**Pull request by gyfora**
This PR contains a few key improvements for savepoint triggering / management:

 1. Unify and simplify savepoint triggering between application and session jobs
 2. Simplify savepoint info update logic
 3. Add trigger type information to savepoints (we can also add new features to the cleanup logic based on this as a next step)
 4. Introduce periodic savepoint triggering
 5. Improve test coverage

### CODE REVIEW

**Comment by gyfora** [comment]
cc @Aitozi @morhidi

**Comment by morhidi** [comment]
Thanks @gyfora looks promising. I'll have a look later today.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 1]
```diff
@@ -163,4 +163,12 @@ public class KubernetesOperatorConfigOptions {
...
                             "Custom HTTP header for HttpArtifactFetcher. The header will be applied when getting the session job artifacts. "
                                     + "Expected format: headerKey1:headerValue1,headerKey2:headerValue2.");
+
+    public static final ConfigOption<Duration> PERIODIC_SAVEPOINT_INTERVAL =
+            ConfigOptions.key("kubernetes.operator.periodic.savepoint.interval")
+                    .durationType()
+                    .noDefaultValue()
+                    .withDescription(
+                            "Interval at which periodic savepoints will be triggered. "
+                                    + "The triggering schedule is not guaranteed an will be done as part of the regular reschedule loop.");
```
Comment (morhidi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:173]: nit: reschedule loop->reconcile loop,  sounds better to me here

File: docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html [hunk 2]
```diff
@@ -74,6 +74,12 @@
             <td>Duration</td>
             <td>The interval before a savepoint trigger attempt is marked as unsuccessful.</td>
         </tr>
+        <tr>
+            <td><h5>kubernetes.operator.periodic.savepoint.interval</h5></td>
```
Comment (morhidi) [inline docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html:78]: Do we expect this property being set globally?

Comment (gyfora) [reply docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html:78]: Users can choose to set it globally or override on a per job level, it works both ways.

Comment (gyfora) [reply docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html:78]: I can change this so setting it to 0 would turn it off explicitly

**Comment by gyfora** [comment]
@morhidi I ran into this issue also with the disposal. I think it's unrelated to my change this is what prompted me to try to improve the dispose logic a little. I will dig a little further into this tomorrow

**Comment by morhidi** [comment]
@gyfora I was comparing the savepoint history with the ones on the dashboard. Does it make sense to display it the same way e.g. latest on top, human readable timestamp?

**Review by Aitozi — COMMENTED** [review]
Thanks @gyfora It looks good, I left some minor comments.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java [hunk 3]
```diff
@@ -653,7 +655,10 @@ public SavepointFetchResult fetchSavepointInfo(
     public void disposeSavepoint(String savepointPath, Configuration conf) throws Exception {
         try (RestClusterClient<String> clusterClient =
                 (RestClusterClient<String>) getClusterClient(conf)) {
-            clusterClient.disposeSavepoint(savepointPath);
+            clusterClient.sendRequest(
```
Comment (Aitozi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java:658]: I check the `disposeSavepoint` internal implementation, why current way is better?

File: docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html [hunk 4]
```diff
@@ -74,6 +74,12 @@
...
+            <td style="word-wrap: break-word;">(none)</td>
+            <td>Duration</td>
+            <td>Interval at which periodic savepoints will be triggered. The triggering schedule is not guaranteed an will be done as part of the regular reschedule loop.</td>
```
[the remark this answers is not in the document]
Comment (gyfora) [reply docs/layouts/shortcodes/generated/kubernetes_operator_config_configuration.html:81]: fixing

**Comment by Sushant20** [comment]
@gyfora Related question: If periodic savepoint is enabled, how do we restore the job with specific savepoint path as part of native integration operator? There is something similar available on other operator page: https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/blob/master/docs/savepoints_guide.md#starting-a-job-from-a-savepoint

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/SavepointUtils.java [hunk 5]
```diff
@@ -17,40 +17,146 @@
...
+                && SavepointUtils.savepointInProgress(jobStatus)) {
+            jobStatus.getSavepointInfo().resetTrigger();
+            LOG.error("Job is not running, cancelling savepoint operation");
+            EventUtils.createOrUpdateEvent(
+                    client,
+                    resource,
+                    EventUtils.Type.Warning,
+                    "SavepointError",
+                    "Savepoint failed for savepointTriggerNonce: "
+                            + resource.getSpec().getJob().getSavepointTriggerNonce(),
```
[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/SavepointUtils.java:158]: 👍

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/validation/DefaultValidator.java [hunk 6]
```diff
@@ -197,6 +197,12 @@ private Optional<String> validateJobSpec(JobSpec job, Map<String, String> confMa
                         String.format(
                                 "Savepoint could not be manually triggered for the running job while config key[%s] is not set",
                                 CheckpointingOptions.SAVEPOINT_DIRECTORY.key()));
+            } else if (configuration.contains(
+                    KubernetesOperatorConfigOptions.PERIODIC_SAVEPOINT_INTERVAL)) {
+                return Optional.of(
```
[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/validation/DefaultValidator.java:202]: They are not that bad :) at least it gives a clear error

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java [hunk 7]
```diff
@@ -640,8 +642,8 @@ public SavepointFetchResult fetchSavepointInfo(
                 }
             }
             Savepoint savepoint =
-                    new Savepoint(
-                            System.currentTimeMillis(), response.get().resource().getLocation());
+                    Savepoint.of(
+                            response.get().resource().getLocation(), SavepointTriggerType.UNKNOWN);
```
[the remark this answers is not in the document]
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java:646]: I will actually change the fetch method to return only the location which is the only thing it actually gets.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java (hunk 3 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java:656]: The disposeSavepoint method of the RestClusterClient, first triggers the dispose action and then keeps polling Flink until it completes.

The way we use it, this is completely unnecessary as we do not handle/retry dispose actions from our side anyways. This way we can save some rest api calls.

However we lose the error logs unfortunately if the dispose actually failed.

**Comment by gyfora** [comment]
@Aitozi I fixed your comments

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java [hunk 8]
```diff
@@ -286,17 +286,15 @@ private static boolean jmMissingForRunningDeployment(FlinkDeployment deployment)
...
     }
 
-    public static boolean isJobRunning(FlinkDeploymentStatus status) {
-        JobManagerDeploymentStatus deploymentStatus = status.getJobManagerDeploymentStatus();
-        return deploymentStatus == JobManagerDeploymentStatus.READY
-                && org.apache.flink.api.common.JobStatus.RUNNING
-                        .name()
-                        .equals(status.getJobStatus().getState());
+    public static boolean isJobRunning(CommonStatus<?> status) {
+        return org.apache.flink.api.common.JobStatus.RUNNING
```
Comment (SteNicholas) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:295]: Could this indeed represent wheter the job is running? IMO, this should check with the task execution states in the job.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/status/SavepointTriggerType.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+/** Savepoint trigger mechanism. */
+public enum SavepointTriggerType {
+    /** Savepoint manually triggered by changing the savepointTriggerNonce. */
+    MANUAL,
+    /** Savepoint periodically triggered by the operator. */
+    PERIODIC,
+    /** Savepoint triggered during stateful upgrade. */
+    UPGRADE,
+    /** Savepoint trigger mechanism unknown. */
+    UNKNOWN
```
Comment (SteNicholas) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/status/SavepointTriggerType.java:29]: Which case is the `UNKNOWN` used for?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java (hunk 8 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/ReconciliationUtils.java:295]: I agree that this should be improved, but completely out of scope for this PR :) 

I will open a follow up jira.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/status/SavepointTriggerType.java (hunk 9 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/status/SavepointTriggerType.java:29]: When the mechanism is not know, for example any savepoint triggered before this update will not have this information.

Or in cases when the savepoint info is retrieved from the flink cluster directly

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/SavepointUtils.java [hunk 10]
```diff
@@ -17,40 +17,153 @@
...
+                                                .getJob()
+                                                .getSavepointTriggerNonce());
+        if (triggerNonceChanged) {
+            return Optional.of(SavepointTriggerType.MANUAL);
+        }
+
+        var savepointInterval =
+                conf.getOptional(KubernetesOperatorConfigOptions.PERIODIC_SAVEPOINT_INTERVAL)
+                        .map(Duration::toMillis)
+                        .orElse(0L);
```
Comment (SteNicholas) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/SavepointUtils.java:116]: Could the default value of `PERIODIC_SAVEPOINT_INTERVAL ` be 0 milliseconds?

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/SavepointUtils.java:116]: makes sense

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/status/SavepointTriggerType.java (hunk 9 above)
Comment (SteNicholas) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/crd/status/SavepointTriggerType.java:29]: @gyfora, are the above cases above mentioned need to add in comments or document to illustrate?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java (hunk 3 above)
Comment (Aitozi) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/service/FlinkService.java:656]: Got it.

**Review by Aitozi — APPROVED** [review]

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Pull request merged**